pulsar/source.go

```diff
@@ -0,0 +1,161 @@
+package pulsar
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	pulsar "github.com/apache/pulsar-client-go/pulsar"
+)
+
+type PulsarSource struct {
+	conf     PulsarConf
+	client   pulsar.Client
+	hook     SourceHook
+	consumer pulsar.Consumer
+}
+
+func (p *PulsarSource) GetKey(msg interface{}) []byte {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetPartition(msg interface{}) int32 {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetValue(msg interface{}) []byte {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) GetOffset(msg interface{}) int64 {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (p *PulsarSource) RegisterHook(hook SourceHook) {
+	p.hook = hook
+}
+
+func GetPulsarSource(conf PulsarConf) *PulsarSource {
+	return &PulsarSource{conf: conf}
+}
+
+// Generate is Source method implementation, which connects to Pulsar and pushes
+// PulsarMessage into the channel
+func (p *PulsarSource) Generate(out chan<- interface{}) {
+	// Prepare KeyFile
+	props := map[string]string{
+		"type":          "client_credentials",
+		"client_id":     p.conf.AuthClientId,
+		"client_secret": p.conf.AuthClientSecret,
+		"issuer_url":    p.conf.AuthIssuerURL,
+	}
+	privateKey, _ := json.Marshal(props)
+	fmt.Println("prepared pulsar privateKey", privateKey)
+
+	// Build authentication
+	auth := pulsar.NewAuthenticationOAuth2(map[string]string{
+		"type":       "client_credentials",
+		"issuerUrl":  p.conf.AuthIssuerURL,
+		"audience":   p.conf.AuthAudience,
+		"privateKey": fmt.Sprintf("data://%s", string(privateKey)),
+		"clientId":   p.conf.AuthClientId,
+	})
+	fmt.Println("prepared pulsar oauth2")
```

Comment:
remove these fmt.Println debug print to stdout

Reply:
ack